### PR TITLE
Added Database for Schedule persistency

### PR DESCRIPTION
**Pull request by lkiesow**
This patch adds SQLAlchemy to store events persistently as well as
enable asynchronous handling of events. This is the first of a series of
patches enabling schedule persistence, an-hoc recordings, recording
history and statistics.

### CODE REVIEW

File: pyca/db.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    end = Column('end', Integer(), nullable=False)
+    uid = Column('uid', String(255), nullable=False)
+    data = Column('data', LargeBinary(), nullable=False)
+    protected = Column('protected', Boolean(), nullable=False, default=False)
+
+    def get_data(self):
+        return json.loads(self.data)
+
+    def set_data(self, data):
+        if sys.version_info[0] == 2:
```
Comment (JanKoppe) [inline pyca/db.py:75]: What's the purpose of this? Is this related to differences between Python 2 and 3? A comment would probably help.

File: pyca/config.py [hunk 2]
```diff
@@ -3,13 +3,18 @@
...
+
+
 __CFG = '''
 [agent]
 name             = string(default='pyca')
 ignore_timezone  = boolean(default=False)
 update_frequency = integer(min=5, default=60)
 cal_lookahead    = integer(min=0, default=14)
 backup_mode      = boolean(default=false)
+database         = string(default='sqlite:///pyca.db')
```
Comment (JanKoppe) [inline pyca/config.py:17]: This doesn't seem to be the same as the default described in `pyca.conf`. Wouldn't this just put the db file in the fs-root?

**Comment by JanKoppe** [comment]
default db string needs to be looked at, other than that: ----

Edit: Woops, need to redact LGTM. Look at the primary key comment, please.

File: pyca/db.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    return Session()
+
+
+# Database Schema Definition
+class Event(Base):
+    '''Database definition of an artist.'''
+
+    __tablename__ = 'event'
+
+    start = Column('start', Integer(), primary_key=True)
```
Comment (JanKoppe) [inline pyca/db.py:59]: Isn't uid better suited as a Primary Key? This way we would have to rely on Opencast to avoid scheduling two recordings at exactly the same start time. Not sure if this is a good thing.

Comment (lkiesow) [reply pyca/db.py:59]: `uid` would be an obvious choice but `start` should be safe as well since we do have a major problem anyway if we have two recordings with one start time (this should fail).

The reason for me choosing the start time is that this could get us in trouble if you do a last-minute change of a recording in Opencast. To give us more time, assume the network is down ;-P
1. Original event time 2pm
2. pyCA gets the schedule
3. Network down (would work with last minute recordings as well, but here it's more obvious)
4. User changes the start time in Opencast to 4pm
5. Network is up again at 3pm, recording is finished
6. pyCA updates the schedule
   - Future events are flushed (not the ones already recorded)
   - pyCA gets changed recording with the same uid
   - _Database exception_

To prevent this you would need to try removing events for each `uid` you get in a new schedule. That would be a lot more work.

Comment (lkiesow) [reply pyca/db.py:59]: I noticed that I failed to properly use the new Github review tool and never submitted the last comment. Sorry ;-(

But then I updated this database patch and changed to SQL schema so that both uid and start are the primary key. I've also added a status column to indicate errors, …

File: pyca/ca.py [hunk 4]
```diff
@@ -164,18 +175,19 @@ def get_schedule():
         logging.error(traceback.format_exc())
         return False
     db = get_session()
-    db.query(Event).filter(Event.end > get_timestamp())\
+    db.query(Event).filter(Event.status == Status.UPCOMMING)\
```
Comment (JanKoppe) [inline pyca/ca.py]: typo. should be UPCOMING

File: pyca/db.py [hunk 5]
```diff
@@ -37,17 +38,30 @@ def get_session():
     return Session()
 
 
+class Status():
+    '''Event status definitions
+    '''
+    UPCOMMING = 1
```
Comment (JanKoppe) [inline pyca/db.py]: typo, should bei UPCOMING

File: pyca/db.py [hunk 6]
```diff
@@ -37,17 +38,30 @@ def get_session():
...
     __tablename__ = 'event'
 
+    uid = Column('uid', String(255), nullable=False, primary_key=True)
     start = Column('start', Integer(), primary_key=True)
     end = Column('end', Integer(), nullable=False)
-    uid = Column('uid', String(255), nullable=False)
     data = Column('data', LargeBinary(), nullable=False)
     protected = Column('protected', Boolean(), nullable=False, default=False)
+    status = Column('status', Integer(), nullable=False,
+                    default=Status.UPCOMMING)
```
Comment (JanKoppe) [inline pyca/db.py]: typo, should be UPCOMING

File: pyca/db.py (hunk 3 above)
Comment (JanKoppe) [reply pyca/db.py:59]: Oh right, didn't think of that scenario, but this would indeed be more likely than double scheduling. good point.

File: pyca/db.py (hunk 6 above)
Comment (lkiesow) [reply pyca/db.py]: UPCOMING it is
